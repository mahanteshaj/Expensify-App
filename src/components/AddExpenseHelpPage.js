import React from 'react'
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'

const AddExpenseHelpPage = () => (
    <div>
        This is form AddExpenseHelpPage component
        </div>
)
export default AddExpenseHelpPage;