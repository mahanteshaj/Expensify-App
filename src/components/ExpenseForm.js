import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

import { fail } from 'assert';
const now = moment()
console.log(now.format('MMM Do,YYYY'))
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            error: '',
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false

        }
    }
    onDescriptionChange = (e) => {
        console.log('sada')
        const description = e.target.value;
        this.setState(() => ({

            description: description
        }))

    }
    onAmountChange = (e) => {
        console.log('Amount has been changed!!!')
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
     }
    onNoteChange = (e) => {
        console.log('note has been changed!!!')
        const note = e.target.value;
        this.setState(() => ({
            note

        }))
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
        //this.setState(()=>({createdAt}))

    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))

    }
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            ///error message 
            this.setState(() => ({

                error: ' Please Provide description and amount'
            }))
        }
        else {
            error: '',
                this.props.onSubmit({
                    description: this.state.description,
                    amount: parseFloat(this.state.amount, 10) * 100,
                    createdAt: this.state.createdAt.valueOf(),
                    note: this.state.note
                })
            console.log('submitted')
        }
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        placeholder="description"
                        type="text"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus
                    />
                    <input
                        placeholder="Amount"
                        type="text"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onFocusChange={this.onFocusChange}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for expense here..."
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>AddExpense</button>
                </form>
                {!!this.state.error && <p>{this.state.error}</p>}

            </div>
        )
    }
}