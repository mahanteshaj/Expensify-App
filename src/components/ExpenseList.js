import React from 'react'
import { connect } from 'react-redux'
import filters from '../reducers/filters';
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (<p>No expenses</p>) : (props.expenses.map((eachExpense) =>

                <ExpenseListItem key={eachExpense.id}{...eachExpense} />
            ))
        }
         </div>
)
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
        //filters:state.filters
    };
}
export default connect(mapStateToProps)(ExpenseList)