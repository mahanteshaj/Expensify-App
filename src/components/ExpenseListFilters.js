import React from 'react'
import { connect } from 'react-redux'
import {DateRangePicker} from 'react-dates'
import { setTextFilter, sortByDate, sortByAmount,setStartDate,setEndDate } from '../actions/filters'
export class ExpenseListFilters extends React.Component{
state={
    calendarFocused:null
}
    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)

    }
    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({
            calendarFocused
        }))         
    }
    onTextChange=(e)=>{
        this.props.setTextFilter(e.target.value)

    }
    onSelectChange = (e) => {
    console.log("clicked")
    e.target.value === 'Date' ? this.props.sortByDate() : this.props.sortByAmount()
}
    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}
                />
                <select onChange={this.onSelectChange}>
                    <option value="Date">Date </option>
                    <option value="Amount">Amount </option>
                </select>
                 <DateRangePicker 
                 startDate={this.props.filters.startDate}
                 endDate={this.props.filters.endDate}
                 onDatesChange={this.onDatesChange}
                 focusedInput={this.state.calendarFocused}
                 onFocusChange={this.onFocusChange}
                 isOutsideRange={()=>false}
                 numberOfMonths={1}
                 showClearDates={true}
                 >
                 </DateRangePicker>   

            </div>

        )
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filters

    }
}
const mapDispatchToProps=(dispatch)=>({
    setTextFilter:(text)=>dispatch(setTextFilter(text)),
    sortByAmount:()=>dispatch(sortByAmount()),
    sortByDate:()=>dispatch(sortByDate()),
    setStartDate:()=>dispatch(setStartDate(startDate)),
    setEndDate:()=>dispatch(setEndDate(endDate))

})
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters); 