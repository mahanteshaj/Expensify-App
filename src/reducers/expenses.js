import axios from 'axios'
const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {

     
    switch (action.type) {
        case 'ADD_EXPENSE': console.log({...action.expense})
        fetch('http://localhost:8080/expensify', { method: 'POST' , headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }, body: JSON.stringify(
        action.expense
        )
        })
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            fetch('http://localhost:8080/expensify/'+action.expense.id,{method:'DELETE'})
            return state.filter(({ id }) => id !== action.expense.id)
        
        case 'EDIT_EXPENSE': fetch('http://localhost:8080/expensify/' + action.id, {
            method: 'PUT', headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }, body: JSON.stringify(
                action.updates
            )
        })
                
            return state.map((expense) => {
                if (expense.id == action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }

                }
                else {
                    return expense
                }
            })
         case 'SORT_BY_AMOUNT': 
                state.sort(function (action1, action2) {
            return action1.amount - action2.amount;

        })
        default:
            return state

    }
}
export default expensesReducer;