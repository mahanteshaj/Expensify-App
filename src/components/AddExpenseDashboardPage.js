import React from 'react'
import ExpenseList from './ExpenseList.js'
import ExpenseListItem from './ExpenseListItem.js';
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'
import ExpenseListFilters from './ExpenseListFilters'
export const ExpenseDashboardPage = () => (<div>
    This is from dashboard component
    <ExpenseListFilters/>
    <ExpenseList />
    <ExpenseListItem/>

</div>
)
export default ExpenseDashboardPage;