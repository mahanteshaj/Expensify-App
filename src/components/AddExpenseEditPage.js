import React from 'react'
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'
import { connect } from 'react-redux'
import ExepenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses';
export class AddExpenseEditPage extends React.Component {
    onSubmit = (expense) => {
        // console.log('updated', expense);
        // console.log('sdsd', this.props)
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')

    }
    onRemove = (expense) => {
        this.props.removeExpense({ id: this.props.expense.id })
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <ExepenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />

                <button onClick={this.onRemove}>Remove </button>

            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => {
            dispatch(editExpense(id, expense))
        },

        removeExpense: (id) => {
            dispatch(removeExpense(id))
        },
    }
}


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(AddExpenseEditPage);