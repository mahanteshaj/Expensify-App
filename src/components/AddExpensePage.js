import React from 'react'
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'
import ExpenseForm from './ExpenseForm'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //  props.dispatch(addExpense(expense))
        this.props.addExpense(expense)
        this.props.history.push('/')
        console.log(expense)
    }
    render() {
        return (
            <div>
                This is form AddExpensePage component
            <h1>Expense Form</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )

    }


}
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => {
            dispatch(addExpense(expense))
        }
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);