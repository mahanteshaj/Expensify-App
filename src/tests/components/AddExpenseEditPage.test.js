import React from 'react'
import {shallow} from 'enzyme'
import { AddExpenseEditPage } from '../../components/AddExpenseEditPage'
import expenses from '../fixtures/expenses'

let editExpense,history,wrapper,removeExpense
beforeEach(() => {
    editExpense = jest.fn()
    removeExpense=jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        <AddExpenseEditPage
            editExpense={editExpense}
            expense={expenses[1]}
            history={history}
            removeExpense={removeExpense}
        />
    )
})
test('test for the adding EXpense page', () => {
    expect(wrapper).toMatchSnapshot()
})
test('should handll edit page with valid input', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')({
        id:expenses[1].id,
        description:'sdsd',
        amount:expenses[1].amount,
        note:expenses[1].note
    })
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,expenses[1])
})
test('should handll remove with valid input', () => {
    wrapper.find('button').prop('onClick')(expenses[1].id)
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[1].id})
})
