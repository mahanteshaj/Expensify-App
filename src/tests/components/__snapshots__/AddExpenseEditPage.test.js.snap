// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for the adding EXpense page 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1312,
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "this  is second description",
        "id": "12224",
        "note": "this is second note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove 
  </button>
</div>
`;
